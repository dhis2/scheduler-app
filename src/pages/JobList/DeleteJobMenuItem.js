import React, { useState, useContext } from 'react'
import { PropTypes } from '@dhis2/prop-types'
import { MenuItem } from '@dhis2/ui'
import i18n from '@dhis2/d2-i18n'
import { DeleteJobModal } from '../../components/Modal'
import { RefetchJobsContext } from '../../components/Context'

const DeleteJobMenuItem = ({ id }) => {
    const [showModal, setShowModal] = useState(false)
    const refetch = useContext(RefetchJobsContext)

    return (
        <React.Fragment>
            <MenuItem
                dense
                destructive
                onClick={() => {
                    setShowModal(true)
                }}
                label={i18n.t('Delete')}
            />
            {showModal && (
                <DeleteJobModal
                    id={id}
                    hideModal={
                        /* istanbul ignore next */
                        () => setShowModal(false)
                    }
                    onSuccess={refetch}
                />
            )}
        </React.Fragment>
    )
}

const { string } = PropTypes

DeleteJobMenuItem.propTypes = {
    id: string.isRequired,
}

export default DeleteJobMenuItem
