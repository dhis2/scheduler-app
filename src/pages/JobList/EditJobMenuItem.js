import React from 'react'
import { string } from '@dhis2/prop-types'
import { MenuItem } from '@dhis2/ui'
import i18n from '@dhis2/d2-i18n'
import history from '../../services/history'

const EditJobMenuItem = ({ id }) => (
    <MenuItem
        dense
        onClick={() => history.push(`/edit/${id}`)}
        label={i18n.t('Edit')}
    />
)

EditJobMenuItem.propTypes = {
    id: string.isRequired,
}

export default EditJobMenuItem
