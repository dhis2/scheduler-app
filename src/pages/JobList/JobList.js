import React from 'react'
import { Card } from '@dhis2/ui-core'
import Title from '../../components/Title'

const JobList = () => (
    <React.Fragment>
        <Title priority={2}>Scheduled Jobs</Title>
        <Card>Some content here</Card>
    </React.Fragment>
)

export default JobList
