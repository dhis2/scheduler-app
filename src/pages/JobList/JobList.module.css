.header {
    display: flex;
    align-items: center;
    margin-bottom: var(--spacers-dp12);
}

.headerTitle {
    font-size: 18px;
    font-weight: 400;
    line-height: 1;
    margin: 0 var(--spacers-dp8) 0 0;
}

.headerLink {
    align-items: center;
    display: flex;
    font-size: 12px;
    color: var(--colors-grey600);
    fill: var(--colors-grey600);
}

.headerLinkIcon {
    margin-right: var(--spacers-dp4);
}

.controlContainer {
    align-items: center;
    display: flex;
    padding: var(--spacers-dp12) var(--spacers-dp12) var(--spacers-dp16);
}

.controlRight {
    display: flex;
    margin-left: auto;
}

.controlRight > * {
    margin-right: var(--spacers-dp24);
}

.controlRight > *:last-child {
    margin-right: 0;
}
