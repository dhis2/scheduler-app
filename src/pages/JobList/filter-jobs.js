const filterJobs = ({ jobFilter, showSystemJobs, jobs }) => {
    // Filter jobs by the current jobFilter
    const applyJobFilter = (job) =>
        job.name.toLowerCase().includes(jobFilter.toLowerCase())

    // Filter jobs depending on the current showSystemJobs
    const applyShowSystemJobs = (job) =>
        // Jobs that are configurable are user jobs
        showSystemJobs ? true : job.configurable

    return jobs.filter(applyJobFilter).filter(applyShowSystemJobs)
}

export default filterJobs
