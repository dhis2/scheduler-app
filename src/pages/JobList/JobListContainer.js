import React, { useEffect } from 'react'
import { object, arrayOf, func, string, bool } from 'prop-types'
import { connect } from 'react-redux'
import { CircularLoader } from '@dhis2/ui-core'
import { actions, selectors } from '../../data/jobs'
import * as rootSelectors from '../../rootSelectors'
import { selectors as entitySelectors } from '../../data/entities'
import JobList from './JobList'

export const UnconnectedJobListContainer = ({
    didFetchSuccessfully,
    isFetching,
    errorMessage,
    jobIds,
    jobEntities,
    fetchJobsIfNeeded,
}) => {
    useEffect(() => {
        fetchJobsIfNeeded()
    }, [fetchJobsIfNeeded])

    if (isFetching && !didFetchSuccessfully) {
        return <CircularLoader />
    }

    if (errorMessage) {
        return <span>{errorMessage}</span>
    }

    return <JobList jobIds={jobIds} jobEntities={jobEntities} />
}

UnconnectedJobListContainer.propTypes = {
    didFetchSuccessfully: bool.isRequired,
    isFetching: bool.isRequired,
    errorMessage: string.isRequired,
    jobIds: arrayOf(string).isRequired,
    jobEntities: object.isRequired,
    fetchJobsIfNeeded: func.isRequired,
}

const mapStateToProps = state => {
    const jobs = rootSelectors.getJobs(state)
    const entities = rootSelectors.getEntities(state)

    return {
        didFetchSuccessfully: selectors.getDidFetchSuccessfully(jobs),
        isFetching: selectors.getIsFetching(jobs),
        errorMessage: selectors.getErrorMessage(jobs),
        jobIds: selectors.getResult(jobs),
        jobEntities: entitySelectors.getJobs(entities),
    }
}

const mapDispatchToProps = {
    fetchJobsIfNeeded: actions.fetchJobsIfNeeded,
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(UnconnectedJobListContainer)
