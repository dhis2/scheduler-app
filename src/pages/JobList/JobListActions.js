import React from 'react'
import { PropTypes } from '@dhis2/prop-types'
import { Menu, DropdownButton } from '@dhis2/ui'
import i18n from '@dhis2/d2-i18n'
import EditJobMenuItem from './EditJobMenuItem'
import RunJobMenuItem from './RunJobMenuItem'
import DeleteJobMenuItem from './DeleteJobMenuItem'

const JobListActions = ({ id }) => (
    <DropdownButton
        small
        component={
            <Menu>
                <EditJobMenuItem id={id} />
                <RunJobMenuItem id={id} />
                <DeleteJobMenuItem id={id} />
            </Menu>
        }
    >
        {i18n.t('Actions')}
    </DropdownButton>
)

const { string } = PropTypes

JobListActions.propTypes = {
    id: string.isRequired,
}

export default JobListActions
