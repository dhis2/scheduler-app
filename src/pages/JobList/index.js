import JobListContainer from './JobListContainer'

export { JobListContainer }
