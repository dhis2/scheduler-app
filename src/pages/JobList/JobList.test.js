import React from 'react'
import { shallow } from 'enzyme'
import { useJobsAndQueues } from '../../hooks/jobs-and-queues'
import JobList from './JobList'

jest.mock('../../hooks/jobs-and-queues', () => ({
    useJobsAndQueues: jest.fn(),
}))

jest.mock('../../components/Store', () => ({
    useJobFilter: jest.fn(() => ['', () => {}]),
    useShowSystemJobs: jest.fn(() => [false, () => {}]),
}))

describe('<JobList>', () => {
    it('renders a spinner when loading the jobs and queues', () => {
        useJobsAndQueues.mockImplementation(() => ({ loading: true }))

        const wrapper = shallow(<JobList />)
        const spinner = wrapper.find('Spinner')

        expect(spinner).toHaveLength(1)
    })

    it('renders errors encountered during fetching', () => {
        useJobsAndQueues.mockImplementation(() => ({
            loading: false,
            error: new Error('Something went wrong'),
        }))

        const wrapper = shallow(<JobList />)
        const noticebox = wrapper.find('NoticeBox')

        expect(noticebox).toHaveLength(1)
    })

    it('renders without errors when loading has completed', () => {
        useJobsAndQueues.mockImplementation(() => ({
            loading: false,
            error: undefined,
            data: [
                {
                    name: '',
                    configurable: true,
                },
            ],
            refetch: () => {},
        }))

        const wrapper = shallow(<JobList />)
        const jobtable = wrapper.find('JobTable')

        expect(jobtable).toHaveLength(1)
    })
})
