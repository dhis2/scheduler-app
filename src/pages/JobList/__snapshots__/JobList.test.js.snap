// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JobList> renders correctly 1`] = `
<Fragment>
  <Title
    priority={2}
  >
    Scheduled Jobs
  </Title>
  <Card>
    Some content here
  </Card>
</Fragment>
`;
